api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -644,6 +646,16 @@ static void checkCompatibility(PartitionSpec spec, Schema schema) {
             "Invalid source type %s for transform: %s",
             sourceType,
             transform);
+        // The only valid parent types for a PartitionField are StructTypes. This must be checked
+        // recursively.
+        int fieldId = field.sourceId();
+        Integer parentId;
+        while ((parentId = parents.get(fieldId)) != null) {
+          Type parentType = schema.findType(parentId);
+          ValidationException.check(
+              parentType.isStructType(), "Invalid partition field parent: %s", parentType);
+          fieldId = parentId;
+        }
```

Comment:
Nit, I think this is easier to read:
```suggestion
        Integer parentId = parents.get(fieldId);
        while (parentId != null) {
          Type parentType = schema.findType(parentId);
          ValidationException.check(
              parentType.isStructType(), "Invalid partition field parent: %s", parentType);
          parentId =  parents.get(parentId);
        }
```